### PR TITLE
feature: experimental paste plugin for raw edtr.io JSON state

### PR DESCRIPTION
**Pull request by elbotho**

### CODE REVIEW

File: src/edtr-io/plugins/paste-hack/editor.tsx [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  const store = useScopedStore()
+
+  function replaceWithStateString() {
+    console.log('starting replace')
+    const newDocuments: DocumentState[] = []
+
+    const content = JSON.parse(dummyContent) as {
+      plugin: 'rows'
+      state: DocumentState[]
+    }
```
Comment (kulla) [inline src/edtr-io/plugins/paste-hack/editor.tsx]: Here you can use io-ts

File: src/edtr-io/plugins/paste-hack/editor.tsx [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    // get parent rows plugin:
+    //const parent = getParent(props.id)(store.getState())
+
+    store.dispatch(
+      replace({
+        id: props.id,
+        plugin: 'rows',
+        state: newDocuments,
+      })
+    )
```
Comment (kulla) [inline src/edtr-io/plugins/paste-hack/editor.tsx]: Here you create a nested row -> we need an implementation where we replace this plugin with multiple other plugins...

Comment (elbotho) [reply src/edtr-io/plugins/paste-hack/editor.tsx]: Yeah, that's one of the questions I asked jonas and then you. Any ideas?

Comment (elbotho) [reply src/edtr-io/plugins/paste-hack/editor.tsx]: https://serlo.slack.com/archives/D0H3UB1E2/p1666600535956429

File: src/edtr-io/plugins/paste-hack/editor.tsx [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  const store = useScopedStore()
+
+  function replaceWithStateString() {
+    console.log('starting replace')
+
+    // TODO: Use io-ts here
+    const content = JSON.parse(dummyContent) as {
+      plugin: 'rows'
+      state: Document[]
+    }
```
Comment (kulla) [inline src/edtr-io/plugins/paste-hack/editor.tsx]: Which types do we want to accept as input?! Examples:

* `type State = { plugin: string, state: unknown }`
* `{ plugin: "rows", state: State[] }`
* `State[]`

File: src/edtr-io/plugins/paste-hack/editor.tsx [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+    const parentPlugin = getParent(props.id)(store.getState())
```
Comment (kulla) [inline src/edtr-io/plugins/paste-hack/editor.tsx]: TODO: Check that parent is rows plugin...

File: src/edtr-io/plugins/paste-hack/editor.tsx [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+      store.dispatch(
+        insertChildBefore({
+          parent: parentPlugin.id,
+          sibling: props.id,
+          document,
+        })
+      )
+    }
+
+    store.dispatch(removeChild({ parent: parentPlugin.id, child: props.id }))
```
Comment (kulla) [inline src/edtr-io/plugins/paste-hack/editor.tsx]: Seems to work. I am fine with deploying it, when it is tested in staging very well. Can we only allow the Paste-Plugin for certain users? (e.g. Anne, Claudia, ...)

**Review by kulla — CHANGES_REQUESTED** [review]

File: src/edtr-io/plugins/paste-hack/editor.tsx (hunk 2 above)
Comment (kulla) [reply src/edtr-io/plugins/paste-hack/editor.tsx]: Here is my idea, which now seems to work: https://github.com/serlo/frontend/pull/1850/commits/a6543689ea29c800008658e27daa8a943a075478

File: src/edtr-io/plugins/paste-hack/editor.tsx (hunk 5 above)
Comment (kulla) [reply src/edtr-io/plugins/paste-hack/editor.tsx]: @elbotho Can you test what happens, when a nested plugin is inserted (e.g. there is a box in the inserted data)

Comment (elbotho) [reply src/edtr-io/plugins/paste-hack/editor.tsx]: > @elbotho Can you test what happens, when a nested plugin is inserted (e.g. there is a box in the inserted data)

Yeah, good idea.

Comment (elbotho) [reply src/edtr-io/plugins/paste-hack/editor.tsx]: > Can we only allow the Paste-Plugin for certain users? (e.g. Anne, Claudia, ...)

I would put the plugin into an experimental feature anyway (this way only power-authors will find it) and we could also only show this experiment for selected roles or users.

File: src/edtr-io/plugins/paste-hack/editor.tsx (hunk 3 above)
Comment (elbotho) [reply src/edtr-io/plugins/paste-hack/editor.tsx]: I tried to make the type more explicit in the io-ts decoder. Basically `{ plugin: "rows", state: State[] }` with a check if the rows child plugins are supported.

File: src/edtr-io/plugins/paste-hack/editor.tsx (hunk 5 above)
Comment (elbotho) [reply src/edtr-io/plugins/paste-hack/editor.tsx]: It's now behind a feature flag and also forbidden in production for now.

File: src/edtr-io/plugins/paste-hack/editor.tsx [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        t.literal('video'),
+        t.literal('table'),
+        t.literal('serloTable'),
+        t.literal('highlight'),
+        t.literal('injection'),
+        t.literal('layout'),
+        t.literal('multimedia'),
+        t.literal('spoiler'),
+        t.literal('important'),
+        t.literal('blockquote'),
```
Comment (kulla) [inline src/edtr-io/plugins/paste-hack/editor.tsx]: We do not support Blockquote any more, right?

File: src/edtr-io/plugins/paste-hack/editor.tsx [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+  plugin: t.literal('rows'),
+  state: t.array(
+    t.strict({
+      plugin: t.union([
+        t.literal('article'),
+        t.literal('articleIntroduction'),
+        t.literal('geogebra'),
+        t.literal('anchor'),
+        t.literal('video'),
+        t.literal('table'),
```
Comment (kulla) [inline src/edtr-io/plugins/paste-hack/editor.tsx]: Same here

File: src/edtr-io/plugins/paste-hack/editor.tsx [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+  const store = useScopedStore()
+  const textareaRef = useRef<HTMLTextAreaElement>(null)
+
+  function throwError(error?: unknown) {
+    showToastNotice('⚠️ Sorry, something is wrong with the data.', 'warning')
+    // eslint-disable-next-line no-console
+    console.error(error)
+    throw new Error(
+      'JSON input data is not a valid edtr-state or contains unsupported plugins'
+    )
```
Comment (kulla) [inline src/edtr-io/plugins/paste-hack/editor.tsx:54]:
```suggestion
    throw new Error(
      'JSON input data is not a valid edtr-state or contains unsupported plugins', { cause: error }
    )
```

Can you check whether it works?

File: src/edtr-io/plugins/paste-hack/editor.tsx [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+export const PasteHackEditor: React.FunctionComponent<PasteHackPluginProps> = (
+  props
+) => {
+  const store = useScopedStore()
+  const textareaRef = useRef<HTMLTextAreaElement>(null)
+
+  function throwError(error?: unknown) {
+    showToastNotice('⚠️ Sorry, something is wrong with the data.', 'warning')
+    // eslint-disable-next-line no-console
+    console.error(error)
```
Comment (kulla) [inline src/edtr-io/plugins/paste-hack/editor.tsx:51]:
```suggestion
```

Maybe not needed when you can add the cause to the error...

File: src/edtr-io/plugins/paste-hack/editor.tsx [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+    throw new Error(
+      'JSON input data is not a valid edtr-state or contains unsupported plugins'
+    )
+  }
+
+  function replaceWithStateString() {
+    if (!textareaRef.current || !textareaRef.current.value) return
+    const decoded = StateDecoder.decode(JSON.parse(textareaRef.current.value))
```
Comment (kulla) [inline src/edtr-io/plugins/paste-hack/editor.tsx]: try-catch needed when content is not JSON parsable...

File: src/edtr-io/plugins/paste-hack/editor.tsx [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+    const content = decoded.right
+
+    try {
+      const parentPlugin = getParent(props.id)(store.getState())
+
+      if (
+        parentPlugin == null ||
+        serializeDocument(parentPlugin.id)(store.getState())?.plugin !== 'rows'
+      ) {
+        const msg = 'only use on rows plugin!'
```
Comment (kulla) [inline src/edtr-io/plugins/paste-hack/editor.tsx]:
```suggestion
        const msg = 'Paste plugin should only be used inside a rows plugin!'
```

File: src/edtr-io/plugins/paste-hack/editor.tsx [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+        showToastNotice(msg)
+        // eslint-disable-next-line no-console
+        console.error(msg)
+        return
+      }
+
+      const documents = content.plugin === 'rows' ? content.state : [content]
+
+      for (const document of documents) {
```
Comment (kulla) [inline src/edtr-io/plugins/paste-hack/editor.tsx]:
```suggestion
      for (const document of content.state) {
```

**Review by kulla — CHANGES_REQUESTED** [review]
@elbotho I think it can go to testing in staging afterwards...

File: src/edtr-io/plugins/paste-hack/editor.tsx (hunk 6 above)
Comment (elbotho) [reply src/edtr-io/plugins/paste-hack/editor.tsx]: in general we still need to support it, but in this case it's probably a good idea to remove 👍

File: src/edtr-io/plugins/paste-hack/editor.tsx (hunk 8 above)
Comment (elbotho) [reply src/edtr-io/plugins/paste-hack/editor.tsx:54]: Can't make it work.
Try e.g. `throw new Error('the error', {cause: "test string"})` in browser console.

**Pull request merged**